README
A collection of warcraft addons developed in Lua

## Buffed
Fully customise the size, location, textures, and timer bars for your buffs, debuffs.
More info at [Wowinterface](http://www.wowinterface.com/downloads/info12613-Buffed.html)
![Alt text](/Buffed/screenshot.png?raw=true "Buffed Screenshot")

## GuildMaster
Automated guild recruitment, donation tracking, alt tracking, and more.
![Alt text](/GuildMaster/screenshot.jpg?raw=true "")

## Keybind
Graphical keybinding system allowing drag and drop assignment for spells and macros.
![Alt text](/GuildMaster/screenshot.jpg?raw=true "")

## KillTheHealer
Monitors the total healing done in a battleground and highlights the top healers when in range, so you can.. remove them.
![Alt text](/KillTheHealer/screenshot.jpg?raw=true "")

## Tournament
Records nearby duels to provide a k/d ratio for players, and an overall rating.
![Alt text](/Tournament/screenshot.jpg?raw=true "")


GuildMaster/Changes.txt

GuildMaster.lua
1.7
	Started moving GM Stats into GMRoster
1.5
	More changes to the way modules are loaded, and the way that the guild frame is loaded.
	Moved some roster functionality over to GMRoster.
1.4+
	Modified some of the function calls to make each module properly modular

GMTimer.lua
2.0 
	-fixed an error preventing the timer repeating.
1.53
	New method for adding timers, makes things slightly neater
	some minor increases in efficiency
1.42
Moved the message queueing function from GMSync to here, which means other addons can still use it when 
GMSync is deactivated.


Modules\GMAlts.lua
2.1
	--removes alts no longer in the guild.

1.4
	Broadcasting alts now works.
	Recieving and recording alts now works.

Modules\GMBanker.lua
2.1
	The guild bank frame should no longer overlap other frames when shown
	away from the guild bank.
	Added a last scanned time.

2.0
	Added button to the guildframe to enable toggling the guild bank
1.6
	Converted to new timer method
	fixed error on checking for auctioneer
	guild bank frame shows away from the guild bank.
1.52
	Created the module - Filters and sorting now work.

Modules\GMEvent.lua


1.47
	-Added option to select minimum resilience, level, and item level.
	-Added a status window showing which players have been invited, etc.
	-Declining an event will automatically decline all events for the next 15 minutes

Modules\GMForum.lua
1.52
fixed bug that hid the  rep bar.

1.4
	Topics can now be completely deleted.
	Guild Masters can delete all posts.

Modules\GMGuildFrame.lua
2.0
	Lots of small fixes to do with the new patch

1.6
	Fixed issue with rewards frame cuasing an error on mouseover for level 1 guilds.

1.5
	Fixed a bug due to the recent interface changes,
	Brought back the info window. This will be used to show/edit the guild information,
	and also show some other stats about the guild.
- 1.47
	Added a button in guild control to temporarily suspend all guild chat.
	Rewards now show properly
	guild masters can delete all posts
	some posts can be properly deleted

Modules\GMLocator.lua
1.52
	Tooltips now show when the map is fullscreen

1.42
	Fixed a problem where guildies that had entered an instance where shown at the last location before entering
	Fixed an error caused by recieving the location of a player in an instance
	

Modules\GMOTD.lua
1.6
	-should sync updates with other players
1.52:
	Hopefully fixed the update bug.
1.5
Should now work!

Modules\GMRanker.lua
1.52:
	merged antirankspam into this module
1.50
	Scrapped the manual vs auto features.
	Integrated the module into the guild control UI ("Rank Requirements" in the guild rank control dropdown)
	fixed a bug where it would still spam per rank change.

Modules\GMRecruit.lua
1.4
	Broadcasting alts now works.
	Recieving and recording alts now works.

Modules\GMRoster.lua
2.0-23
	bug fix:
		search box iis constrain by the show offline button
		resizing the roster columns no longer changes sort order
		class names in the detail frame start with a capital letter.
		added item level
2.0
	Created the new roster
	added talent sharing
1.5
	- moved the roster update functionality into this module
1.4 - CreateModule

Modules\GMSchedule.lua
1.52 Created the addon

Modules\GMSettings.lua
1.52:
	Each module now stores its own settings. This module loads other modules
	settings and displays them in the interface options menu.
1.40:
	Made each module properly modular, added toggling of each module.

Modules\GMStats.lua
	1.4
	Total donations now includes donations from looting.
	Mousewheel to scroll
	escape closes the window.

Modules\GMSync.lua
1.42
	-	Moved the queue function over to GMTimer.
		Added a method of blocking users whose addons are out of date.

Modules\GMTutorial.lua
1.52
	Created the module.


GuildMaster/GuildMaster.lua
--[[
====================================================================================
=========================== GUILD MASTER ============================================
====================================================================================

1. About
2. Changes
3. To do.

--========================== 1. About ===============================
	This module does some database management and handles loading all of the other modules.
	
	Aim of the addon:
	The key aims of this addon are to make guilds more streamlined: 
	Running a guild requires a fair emount of administration, so any way this can be made easier for the guild master is 
	obviously a bonus.
	
	
--========================== 2. Changes ==============================
1.7
	Started moving GM Stats into GMRoster
1.5
	More changes to the way modules are loaded, and the way that the guild frame is loaded.
	Moved some roster functionality over to GMRoster.
1.4+
	Modified some of the function calls to make each module properly modular
--========================== 3. To do ================================



--------------------------------

Each time the player logs in, broadcast some stats, eg:
resi
rating
itemlevel

Take a look at moving some of the savedvariables away from per character, 
eg the database and forum


Make all modules use the stored playername instead of UnitName("player")


current layout: 

GMLogs = {	} -- stores all of the stats per member

GMSettings = {
	Ranklist = {} stores the rank settings
	Debug = boolean -- wether to print debugging stuff or not
	Active -- wether to directly to record to the current table or not.
	LastLogout = number -- last date of the logout
	MOTD -- the guild message of the day, just in case it gets deleted.
	modules = {} --list of  addons, not quite sure what it does.
	Log = {} --list of all guild members with some stats for them.
	Version = number -- the version number of the guild
	EnableEvent = boolean -- wether to enable event insites or not.
	GMRecruit = {
							lastsent = number -- time of last message
							list == table -- list of all messages.
							}
	StartTime ==number when they first come online.
	lockout = boolean --locks the player out if the version number isnt newest
	
	
	
GMForum = {} -- list of all forum posts.

GMBanker = {} - Filters for bank tabs and 

addon 
String - Addon name to look up metadata for
field 
value = GetAddOnMetadata("addon", "field")
String - Field name. May be Title, Notes, Author, Version, or anything starting with X-
}
]]

--[[
	Standard checks to be carried out before release:
	
		-Check no guild info is wiped.
		-Check on multiple guilds
		-Check on character not in a guild.
		-Test upgrade from old svaedvars
		-Test upgrade from no savedvars.
		
	Module checks:
		-Load/Unload each module on demand.

--]]


local ADDON_VERSION

local runcount = 3 --records if the guild roster has updated for the first time.
local firstrun = true

GMaster = {} --stores info on which modules are loaded
GMaster.PlayerName = UnitName("player")
GMaster.QueueTable = {} -- set up the table
GMaster.TempTable = {}
GMaster.loaded = {}

GMaster.LoadOrder = {} --each module puts its name in here to indicate that it needs to be loaded
-- common events 
GMaster.PEW = {} --list of functions to run on PEW
GMaster.AL = {} --list of functions to run on Addon Loaded
GMaster.CMA = {} --list of functions to run on CHAT_MESSAGE_ADDON
GMaster.GFL = {} --guild frame loaded


GMaster.ModuleLoad = {} --stores all methods that load the module
GMaster.ModuleRemove = {} --stores all methods that remove the module
GMaster.ModuleSettings = {} --stores the settings page for each module

GMaster.GUILD_FRAME_LOADED = nil -- has the guild frame loaded yet.
GMaster.IS_GUILD_MASTER = nil

GMSettings = {} --saved vars, stores settings for all modules
GMRoster = {} -- saved vars, stores info for each player.

local SETTINGS_LOADED = nil

GMaster.Debug = function(self, ...) --global print used by all modules to debug
	if GMSettings.Debug then
		print(...)
	end
end

 

function GMaster.EventHandler(self, event, ...)
	if event == "ADDON_LOADED" then
		local addon = ...
		if addon == "GuildMaster" then --if this addon has loaded
			GMaster.OnLoad() --check settings etc
			for name, v in pairs(GMaster.LoadOrder) do
				if GMSettings.modules[name] and GMSettings.modules[name].active and GMaster.AL[name] then --if the module needs to be loaded, load it
					GMaster.AL[name]() --load the module
					GMaster.loaded[name]  = true --store that the module has loaded successfully
				end
			end
		elseif addon == "Blizzard_GuildUI" then
			GMaster.GUILD_FRAME_LOADED = true
			LoadAddOn("Blizzard_GuildControlUI")
			for name, func in pairs(GMaster.GFL) do
				if type(func) == "function" and GMSettings.modules[name] then
					func()
					GMaster.Debug("GFL, ",name, func)
				end
			end
		end
	elseif event == "PLAYER_ENTERING_WORLD" then 
			GMaster.IS_GUILD_MASTER = IsGuildLeader("player")
			for name, func in pairs(GMaster.PEW) do
				if type(func) == "function" then
					func()
				end
				GMaster.Debug("PEW, ",name, func)
			end
			if not SETTINGS_LOADED then
				if GMaster.GMSettings then --this needs to load after all the other modules
					GMaster.GMSettings()
					SETTINGS_LOADED = true
				end
			end		
	elseif event == "CHAT_MSG_ADDON" then
		if IsInGuild() then
			for name, func in pairs(GMaster.CMA) do --loop through all the functions that need th event
				if type(func) == "function" then
					func(...)
				end
			end				
		end
	elseif event == "GUILD_ROSTER_UPDATE" then
		if GMaster.RosterUpdate then
			GMaster.RosterUpdate(...)
		end	
	end
end

function GMaster.OnLoad() 
		ADDON_VERSION = tonumber(GetAddOnMetadata("GuildMaster", "Version"))
		GMaster.LastLogin = time()
		if not GMSettings.Version then
		
				GMSettings = {Version = ADDON_VERSION, StartTime = 0, LastLogout = 0, GMForum = {}, EnableEvent = true}
				GMSettings.modules = {}
				for name, k in pairs(GMaster.LoadOrder) do
					GMSettings.modules[name] = {}
					GMSettings.modules[name].active = true
				end
				GMSettings.RankList = {}
				GMaster.TempTable = {}
				GMSettings.Active = nil
				GMLogs = {}
				GMForum = {}
				
				if not GMRoster then
					GMRoster = {}
				end
		
		end
		
	
	if  GMSettings.Version <  ADDON_VERSION then --if they are using an old version of the addon, 
	
			GMSettings = {Version = ADDON_VERSION, StartTime = 0, LastLogout = 0, GMForum = {}, EnableEvent = true}
			GMSettings.modules = {}
			if not GMSettings.modules then
				GMSettings.modules = {}
			
			end
			for name, k in pairs(GMaster.LoadOrder) do
				GMSettings.modules[name] = {}
				GMSettings.modules[name].active = true
			end
			
			GMSettings.RankList = {}
			GMaster.TempTable = {}
			GMSettings.Active = nil
			GMLogs = {}
			GMForum = {}
			
			if not GMRoster then
				GMRoster = {}
			end
			
			if not GMSettings.RankList then
				GMSettings.RankList = {}
			end
			GMSettings.Version = ADDON_VERSION
	else
		GMSettings.lockout = false --innocent until proven guilty
	end
end



--[[	This big 'ol function needs to clean up our tables, which means
-remove any tables that are teeny tiny
-remove any tables that are empty
-merge tables that are old

--]]
function GMaster.CleanTables()
	
	local temptable = {}
				
	for dated, players in pairs(GMLogs) do --clears empty and small tables
		if not players.LastRecord then --if it hasnt recorded any values
			GMLogs[dated] = nil --remove it
		else
			local wipe = false
			if players.LastRecord - dated < 600 then --remove entries smaller than 10 minutes
				wipe = true
			end
			if wipe then
				GMLogs[dated] = nil --remove it
			else
				tinsert(temptable, {dated, players.LastRecord}) --add it into the temp table
			end
		end
	end
	
	table.sort(temptable, function(a, b) return a[1] > b[1] end)
			
	for i = 2, #temptable, 1 do -- this function should merge tabls that are close to each other
		local first = temptable[i]
		local second = temptable[i +1]
		if first and second then
		
			 --dont add the second to last table to the potentially current table
				if first[1] - second[2] < 600 and (first[1] - second[2] > 0) then
				
				GMLogs[second[1]].LastRecord = first[2]
				--check they are on the same day
					if date("%d", first[1]) == date("%d", second[2]) then 
						for name, info in pairs(GMLogs[first[1]]) do
							if name ~= "LastRecord" then
								if not GMLogs[second[1]][name] then
									GMLogs[second[1]][name] = {}
								end
								for k, v in pairs(info) do
									if not GMLogs[second[1]][name][k] then
										GMLogs[second[1]][name][k] = v
									else
										if k > 9 and k < 16 then
											if  GMLogs[second[1]][name][k] < v then
												GMLogs[second[1]][name][k] = v
											end
										else
											GMLogs[second[1]][name][k] = GMLogs[second[1]][name][k] + v
										end
									end
								end
							end
						end
					end
						GMLogs[first[1]] = nil
				end		
			end
		end
				
	for i = 2, #temptable do
		local first = temptable[i]
		for j = 2, #temptable do
			if i ~= j then
				local second = temptable[j]
				if first[1] > second[1] and first[2] > second[2] then --first was somehow recorded inside second
					GMLogs[first[1]] = nil -- remove it from the savedvars
				end
			
			end
		end
	end
	GMaster.CleanTables = nil
	--need to loop through all tables and merge some based on their dates
	--pull out the merge function from the above loop so it can be reused
end

	--[[
	Need to make this function delete any location data stored when they go offline.
	
	--]]


local Eventler = CreateFrame("Frame")
Eventler:RegisterEvent("PLAYER_ENTERING_WORLD")
Eventler:RegisterEvent("ADDON_LOADED")
Eventler:RegisterEvent("CHAT_MSG_ADDON")
Eventler:RegisterEvent("GUILD_ROSTER_UPDATE")
Eventler:SetScript("OnEvent", GMaster.EventHandler)